File: src/loom/scan.py
"""Filesystem scan: walk markdown files and turn them into IndexRecords.

Used by :mod:`loom.rebuild`, :mod:`loom.validation`, and Phase 3's
single-item sync.
"""

from __future__ import annotations

import hashlib
from collections.abc import Iterator
from datetime import UTC, datetime
from pathlib import Path
from typing import Any

from .ids import (
    ARCHIVE_DIRNAME,
    PROJECTS_DIRNAME,
    ItemType,
    qid_from_path,
)
from .index import IndexRecord, frontmatter_to_json
from .storage import load


def walk_md_files(root: Path) -> Iterator[Path]:
    """Yield every ``.md`` file under ``projects/`` and ``_archive/projects/``.

    Anything outside those subtrees (e.g. files in ``.loom/``, top-level
    READMEs) is ignored. Within those subtrees a non-canonical file
    layout will be reported as a stray-file issue by callers, not by
    this walker.
    """
    candidates = (
        root / PROJECTS_DIRNAME,
        root / ARCHIVE_DIRNAME / PROJECTS_DIRNAME,
    )
    for base in candidates:
        if base.exists():
            yield from sorted(base.rglob("*.md"))


def hash_file_bytes(path: Path) -> str:
    """sha256 of the raw on-disk bytes.

    Hashing the raw bytes — never the parsed-and-rerendered form — keeps
    drift detection honest: ruamel may normalize quoting or spacing on
    rewrite, which would be spurious drift if we hashed parsed content.
    """
    return hashlib.sha256(path.read_bytes()).hexdigest()


def build_record(path: Path, root: Path) -> IndexRecord:
    """Construct an :class:`IndexRecord` from a markdown file.

    Raises :class:`InvalidQualifiedId` if the path doesn't match the
    canonical layout, or :class:`ValidationError` if the file is
    unreadable or its frontmatter parses to something unusable.

    Missing optional fields fall back to sensible defaults; missing
    timestamps fall back to the file's mtime; a missing ``status`` on a
    non-project item defaults to ``"ready"`` (the canonical initial
    state). The path's qid always wins over any disagreeing
    ``qualified_id`` / ``id`` in the frontmatter — those mismatches are
    surfaced as validation issues by the caller, not corrected here.
    """
    qid, archived = qid_from_path(path, root)
    fm, _body = load(path)

    rel_path = str(path.relative_to(root))
    body_hash = hash_file_bytes(path)

    mtime_iso = _path_mtime_iso(path)
    created_at = _opt_str(fm.get("created_at")) or mtime_iso
    updated_at = _opt_str(fm.get("updated_at")) or mtime_iso

    title = _opt_str(fm.get("title")) or f"(untitled) {qid}"

    status: str | None = None
    if qid.type != ItemType.PROJECT:
        status = _opt_str(fm.get("status")) or "ready"

    parent = qid.parent
    parent_id = str(parent) if parent else None

    is_project = qid.type == ItemType.PROJECT
    repo = _opt_str(fm.get("repo")) if is_project else None
    default_branch = _opt_str(fm.get("default_branch")) if is_project else None

    depends_on = tuple(str(x) for x in (fm.get("depends_on") or ()))
    tags = tuple(str(x) for x in (fm.get("tags") or ()))

    return IndexRecord(
        qualified_id=str(qid),
        type=qid.type.value,
        project=qid.project,
        epic=qid.epic,
        story=qid.story,
        task=qid.task,
        parent_id=parent_id,
        title=title,
        status=status,
        assignee=_opt_str(fm.get("assignee")),
        branch=_opt_str(fm.get("branch")),
        pr_url=_opt_str(fm.get("pr_url")),
        repo=repo,
        default_branch=default_branch,
        archived=archived,
        created_at=created_at,
        updated_at=updated_at,
        file_path=rel_path,
        body_hash=body_hash,
        frontmatter_json=frontmatter_to_json(fm),
        depends_on=depends_on,
        tags=tags,
    )


def _opt_str(v: Any) -> str | None:
    if v is None:
        return None
    s = str(v)
    return s if s else None


def _path_mtime_iso(path: Path) -> str:
    return datetime.fromtimestamp(path.stat().st_mtime, UTC).isoformat(timespec="seconds")

File: web/lib/scan.ts
/**
 * Filesystem scan: walk markdown files and turn them into IndexRecords.
 *
 * Mirrors src/loom/scan.py.
 * Used by rebuild.ts and validation.ts.
 */

import * as fs from "fs";
import * as path from "path";
import { load } from "./storage";
import { qid_from_path, ItemType, ARCHIVE_DIRNAME, PROJECTS_DIRNAME } from "./ids";
import { IndexRecord, frontmatterToJson } from "./index";

// ---------------------------------------------------------------------------
// walkMdFiles
// ---------------------------------------------------------------------------

/**
 * Return every .md file path under projects/ and _archive/projects/ in sorted
 * order. Files outside those subtrees are ignored.
 */
export function walkMdFiles(root: string): string[] {
  const candidates = [
    path.join(root, PROJECTS_DIRNAME),
    path.join(root, ARCHIVE_DIRNAME, PROJECTS_DIRNAME),
  ];
  const result: string[] = [];
  for (const base of candidates) {
    if (fs.existsSync(base)) {
      result.push(...collectMdFiles(base));
    }
  }
  return result;
}

function collectMdFiles(dir: string): string[] {
  const results: string[] = [];
  const entries = fs.readdirSync(dir, { withFileTypes: true });
  // Sort for deterministic ordering (mirrors Python's sorted() on rglob)
  entries.sort((a, b) => (a.name < b.name ? -1 : a.name > b.name ? 1 : 0));
  for (const entry of entries) {
    const full = path.join(dir, entry.name);
    if (entry.isDirectory()) {
      results.push(...collectMdFiles(full));
    } else if (entry.isFile() && entry.name.endsWith(".md")) {
      results.push(full);
    }
  }
  return results;
}

// ---------------------------------------------------------------------------
// hashFileBytes
// ---------------------------------------------------------------------------

/**
 * Return the sha256 hex digest of the raw on-disk bytes of filePath.
 *
 * Hashing the raw bytes (never the parsed-and-rerendered form) keeps
 * drift detection honest: an idempotent re-render would be spurious drift.
 */
export async function hashFileBytes(filePath: string): Promise<string> {
  const bytes = fs.readFileSync(filePath);
  const hashBuf = await crypto.subtle.digest("SHA-256", bytes);
  return Array.from(new Uint8Array(hashBuf))
    .map((b) => b.toString(16).padStart(2, "0"))
    .join("");
}

// ---------------------------------------------------------------------------
// buildRecord
// ---------------------------------------------------------------------------

/**
 * Construct an IndexRecord from a markdown file.
 *
 * Path wins over frontmatter: the qid is always derived from the path.
 * Missing optional fields fall back to sensible defaults:
 *   - missing status on a non-project defaults to "ready"
 *   - missing timestamps fall back to the file's mtime
 * repo / default_branch are only read for project items.
 */
export async function buildRecord(
  filePath: string,
  root: string
): Promise<IndexRecord> {
  const [qid, archived] = qid_from_path(filePath, root);
  const [fm] = load(filePath);

  const relPath = path.relative(root, filePath);
  const bodyHash = await hashFileBytes(filePath);

  const mtimeIso = pathMtimeIso(filePath);
  const created_at = optStr(fm["created_at"]) ?? mtimeIso;
  const updated_at = optStr(fm["updated_at"]) ?? mtimeIso;

  const title = optStr(fm["title"]) ?? `(untitled) ${qid.toString()}`;

  let status: string | null = null;
  if (qid.type !== ItemType.PROJECT) {
    status = optStr(fm["status"]) ?? "ready";
  }

  const parent = qid.parent;
  const parent_id = parent ? parent.toString() : null;

  const isProject = qid.type === ItemType.PROJECT;
  const repo = isProject ? optStr(fm["repo"]) ?? null : null;
  const default_branch = isProject ? optStr(fm["default_branch"]) ?? null : null;

  const rawDeps = fm["depends_on"];
  const depends_on = Array.isArray(rawDeps) ? rawDeps.map((x) => String(x)) : [];

  const rawTags = fm["tags"];
  const tags = Array.isArray(rawTags) ? rawTags.map((x) => String(x)) : [];

  return {
    qualified_id: qid.toString(),
    type: qid.type,
    project: qid.project,
    epic: qid.epic,
    story: qid.story,
    task: qid.task,
    parent_id,
    title,
    status,
    assignee: optStr(fm["assignee"]) ?? null,
    branch: optStr(fm["branch"]) ?? null,
    pr_url: optStr(fm["pr_url"]) ?? null,
    repo,
    default_branch,
    archived,
    created_at,
    updated_at,
    file_path: relPath,
    body_hash: bodyHash,
    frontmatter_json: frontmatterToJson(fm as Record<string, unknown>),
    depends_on,
    tags,
  };
}

// ---------------------------------------------------------------------------
// Internal helpers
// ---------------------------------------------------------------------------

function optStr(v: unknown): string | null {
  if (v === null || v === undefined) return null;
  const s = String(v);
  return s || null;
}

function pathMtimeIso(filePath: string): string {
  const mtimeMs = fs.statSync(filePath).mtimeMs;
  return new Date(mtimeMs).toISOString().replace(/\.\d{3}Z$/, "+00:00");
}
